### DISCUSSION TITLE
An error should be shown if code intended for the server side (used in `getServerSideProps` get bundled to the client)

### DISCUSSION THREAD

**Discussion by hgezim** [vercel/next.js#25938] [Ideas]
### Describe the feature you'd like to request

When it comes to `getServerSideProps`, the [docs say](https://nextjs.org/docs/basic-features/data-fetching#write-server-side-code-directly):

> Note that getStaticProps runs only on the server-side. It will never be run on the client-side. It won’t even be included in the JS bundle for the browser. That means you can write code such as direct database queries without them being sent to browsers. 

The [docs also explain](https://nextjs.org/docs/basic-features/data-fetching#getstaticprops-static-generation):

> You can import modules in top-level scope for use in getStaticProps. Imports used in getStaticProps will not be bundled for the client-side.

However, that's not quite true:

![CleanShot 2021-06-09 at 07 09 01](https://user-images.githubusercontent.com/124675/121360370-96f14580-c8f1-11eb-941b-60fa4ed27ccb.png)

In the [code elimination demo](https://next-code-elimination.vercel.app/), we have this code:

```
// This app shows what Next.js bundles for the client-side with its new SSG
// support. This editor supports TypeScript syntax.
import Cookies from 'cookies';
import Mysql from 'mysql';
import Link from 'next/link';
import SQL from 'sql-template-strings';
import Layout from '../components/Layout';

const pool = Mysql.createPool(process.env.DATABASE_URL);

export default function ({ projects }) {
  return (
    <Layout>
      <h1>Projects</h1>
      <ul>
        {projects.map((project) => (
          <li key={project.id}>
            <Link href="/projects/[id]" as={`/projects/${project.id}`}>
              <a>{project.name}</a>
            </Link>
          </li>
        ))}
      </ul>
    </Layout>
  );
}

export async function getServerSideProps({ req, res }) {
  const userId = new Cookies(req, res).get('user_id');
  const projects = await new Promise((resolve, reject) =>
    pool.query(
      SQL`SELECT id, name FROM projects WHERE user_id = ${userId};`,
      (err, results) => (err ? reject(err) : resolve(results))
    )
  );
  return { props: { projects } };
}
```

It's not hard to imagine someone hard typing the connection string to the pool assignment call:

`const pool = Mysql.createPool(`postgres:user:supersecretpassword`);`

And then someone else down the line doing a console.log of that `pool` in the page itself resulting in the page bundling the secret.

### Describe the solution you'd like

An error or warning at some point in time when the compiler detects code being used `getServerSideProps` that's also getting bundled to the client.

### Describe alternatives you've considered

A linter solution? Not sure how hard this would be, honestly. It does scare me that someone down the line not knowing could open up major security vulnerabilities.

**Comment by tahonaPL** [comment]
I have similar have you found some workaround.  https://github.com/vercel/next.js/discussions/32874
